main/src/util/profiler.hpp

```diff
@@ -0,0 +1,147 @@
+        if (_rank == 0) profilingFile.open(filename);
+    }
+    ~Profiler() {}
+
+    void printMetrics(int iteration)
+    {
+        profilingFile << meanPerStep.at(iteration) << "," << stdevPerStep.at(iteration) << ","
+                      << covPerStep.at(iteration) << "," << I_2PerStep.at(iteration) << ","
+                      << lambdaPerStep.at(iteration);
+        for (int i = 0; i < _numRanks; i++)
+        {
+            profilingFile << "," << vectorMetric.at(iteration)[i];
+        }
+    }
+
+    void printProfilingInfo()
+    {
+        if (_rank == 0)
+        {
+            size_t iter = 1;
+            // profilingFile << std::setw(3) << " ";
+            for (int i = 0; i < _numRanks; i++)
+            {
+                profilingFile << "RANK" << i << ",";
+            }
+            profilingFile << "mean,"
+                          << "stdev,"
+                          << "CoV,"
+                          << "I_2,"
+                          << "lambda";
+            for (int i = 0; i < _numRanks; i++)
+            {
+                profilingFile << ",Vector" << i;
+            }
+            profilingFile << std::endl;
+            for (auto& element : timeSteps)
+            {
+                calculateMetrics(element);
+                // profilingFile << std::setw(3) << iter << " ";
+                for (int i = 0; i < _numRanks; i++)
+                {
+                    profilingFile << element[i] << ",";
+                }
+                printMetrics(iter - 1);
+                iter++;
+                profilingFile << std::endl;
+            }
+            profilingFile << std::endl;
+            std::cout << "Profiling data written." << std::endl;
+        }
+        profilingFile.close();
+    }
+
+    void saveTimings(float duration) { funcTimes.push_back(duration); }
+
+    void saveTimestepAndParticleNo(float duration, size_t numLocalParticles)
+    {
+        funcTimes.push_back(duration);                 // save the total timestep timing
+        funcTimes.push_back((float)numLocalParticles); // save the local number of particles
+    }
+
+    void gatherTimings()
+    {
+        std::cout << "Gathering profiling data." << std::endl;
+
+        float* durations = new float[_numRanks * funcTimes.size()];
+        MPI_Gather(funcTimes.data(), funcTimes.size(), MPI_FLOAT, &durations[_rank * funcTimes.size()],
+                   funcTimes.size(), MPI_FLOAT, 0, MPI_COMM_WORLD);
+
+        if (_rank == 0)
+        {
+            for (int i = 0; i < funcTimes.size(); i++)
+            {
+                float* funcDurations = new float[_numRanks];
+                for (int r = 0; r < _numRanks; r++)
+                {
+                    funcDurations[r] = durations[r * funcTimes.size() + i];
+                }
+                timeSteps.push_back(funcDurations);
+            }
+        }
+    }
+
+    // LiB metrics calculations
+    void calculateMetrics(float* durations)
+    {
```

Comment:
This analysis part could be done in Python. I believe it would be more flexible and make it easier to adapt to changing needs in the future.